src/corehost/cli/apphost/apphost.windows.cpp

```diff
@@ -0,0 +1,152 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "apphost.windows.h"
+#include "error_codes.h"
+#include "pal.h"
+#include "trace.h"
+#include "utils.h"
+
+#include <regex>
+
+namespace
+{
+    pal::string_t g_buffered_errors;
+
+    void buffering_trace_writer(const pal::char_t* message)
+    {
+        // Add to buffer for later use.
+        g_buffered_errors.append(message).append(_X("\n"));
+        // Also write to stderr immediately
+        pal::err_fputs(message);
+    }
+
+    // Determines if the current module (apphost executable) is marked as a Windows GUI application
+    bool is_gui_application()
+    {
+        HMODULE module = ::GetModuleHandleW(nullptr);
+        assert(module != nullptr);
+
+        // https://docs.microsoft.com/en-us/windows/win32/debug/pe-format
+        BYTE *bytes = reinterpret_cast<BYTE *>(module);
+        UINT32 pe_header_offset = reinterpret_cast<IMAGE_DOS_HEADER *>(bytes)->e_lfanew;
+        UINT16 subsystem = reinterpret_cast<IMAGE_NT_HEADERS *>(bytes + pe_header_offset)->OptionalHeader.Subsystem;
+
+        return subsystem == IMAGE_SUBSYSTEM_WINDOWS_GUI;
+    }
+
+    void write_errors_to_event_log(const pal::char_t *executable_path, const pal::char_t *executable_name)
+    {
+        // Report errors to the Windows Event Log.
+        auto eventSource = ::RegisterEventSourceW(nullptr, _X(".NET Runtime"));
+        const DWORD traceErrorID = 1023; // Matches CoreCLR ERT_UnmanagedFailFast
+        pal::string_t message;
+        message.append(_X("Description: A .NET Core application failed.\n"));
+        message.append(_X("Application: ")).append(executable_name).append(_X("\n"));
+        message.append(_X("Path: ")).append(executable_path).append(_X("\n"));
+        message.append(_X("Message: ")).append(g_buffered_errors).append(_X("\n"));
+
+        LPCWSTR messages[] = {message.c_str()};
+        ::ReportEventW(eventSource, EVENTLOG_ERROR_TYPE, 0, traceErrorID, nullptr, 1, 0, messages, nullptr);
+        ::DeregisterEventSource(eventSource);
+    }
+
+    void show_error_dialog(const pal::char_t *executable_name, int error_code)
+    {
+        // Show message dialog for UI apps with actionable errors
+        if (error_code != StatusCode::CoreHostLibMissingFailure  // missing hostfxr
+            && error_code != StatusCode::FrameworkMissingFailure) // missing framework
+            return;
+
+        pal::string_t gui_errors_disabled;
+        if (pal::getenv(_X("DOTNET_DISABLE_GUI_ERRORS"), &gui_errors_disabled) && pal::xtoi(gui_errors_disabled.c_str()) == 1)
+            return;
+
+        pal::string_t dialogMsg = _X("To run this application, you must install .NET Core.\n\n");
+        pal::string_t url = DOTNET_CORE_DOWNLOAD_URL;
+        if (error_code == StatusCode::CoreHostLibMissingFailure)
+        {
+            url.append(_X("?runtime=latest"));
```

Comment:
I think this should be more specific - something like `?noruntime`. I would leave it up to the server to decide which version to get (whether latest or something else).

In 5.0 we will have to stamp the apphost with some version information. I would actually add both the TFM as well as the detailed version of apphost itself (which currently is not stored anywhere in the binary, unfortunately, which will make debugging crash dumps quite a bit painful). I created #8525 for this.

Reply:
Changed to `missing_runtime=true`